cpp/include/cudf/utilities/traits.hpp

```diff
@@ -591,63 +592,59 @@ constexpr inline bool is_nested(data_type type)
-  struct is_logically_castable_impl<Type2, Type1> : std::true_type { \
-  };
-#endif
-
-// Allow cast between timestamp and integer representation
-MAP_CASTABLE_TYPES(cudf::timestamp_D, cudf::timestamp_D::duration::rep);
-MAP_CASTABLE_TYPES(cudf::timestamp_s, cudf::timestamp_s::duration::rep);
-MAP_CASTABLE_TYPES(cudf::timestamp_ms, cudf::timestamp_ms::duration::rep);
-MAP_CASTABLE_TYPES(cudf::timestamp_us, cudf::timestamp_us::duration::rep);
-MAP_CASTABLE_TYPES(cudf::timestamp_ns, cudf::timestamp_ns::duration::rep);
-// Allow cast between durations and integer representation
-MAP_CASTABLE_TYPES(cudf::duration_D, cudf::duration_D::rep);
-MAP_CASTABLE_TYPES(cudf::duration_s, cudf::duration_s::rep);
-MAP_CASTABLE_TYPES(cudf::duration_ms, cudf::duration_ms::rep);
-MAP_CASTABLE_TYPES(cudf::duration_us, cudf::duration_us::rep);
-MAP_CASTABLE_TYPES(cudf::duration_ns, cudf::duration_ns::rep);
-// Allow cast between decimals and integer representation
-MAP_CASTABLE_TYPES(numeric::decimal32, numeric::decimal32::rep);
-MAP_CASTABLE_TYPES(numeric::decimal64, numeric::decimal64::rep);
-
 template <typename FromType>
 struct is_logically_castable_to_impl {
-  template <typename ToType>
-  constexpr bool operator()()
+  template <typename ToType,
+            typename std::enable_if_t<not cuda::std::is_trivially_copyable_v<FromType> ||
+                                      not cuda::std::is_trivially_copyable_v<ToType>>* = nullptr>
+  constexpr bool operator()(size_t)
+  {
+    return false;
+  }
+  template <typename ToType,
+            typename std::enable_if_t<(cuda::std::is_trivially_copyable_v<FromType> &&
+                                       cuda::std::is_trivially_copyable_v<
+                                         ToType>)&&not cudf::is_fixed_point<ToType>()>* = nullptr>
+  constexpr bool operator()(size_t fromSize)
   {
-    return is_logically_castable_impl<FromType, ToType>::value;
+    constexpr auto toSize = sizeof(ToType);
+    return fromSize == toSize;
+  }
+
+  template <typename ToType,
+            typename std::enable_if_t<(cuda::std::is_trivially_copyable_v<FromType> &&
+                                       cuda::std::is_trivially_copyable_v<
+                                         ToType>)&&cudf::is_fixed_point<ToType>()>* = nullptr>
+  constexpr bool operator()(size_t fromSize)
+  {
+    constexpr auto toSize = sizeof(typename ToType::rep);
+    return fromSize == toSize;
   }
 };
 
 struct is_logically_castable_from_impl {
-  template <typename FromType>
+  template <typename FromType,
+            typename std::enable_if_t<not cudf::is_fixed_point<FromType>()>* = nullptr>
+  constexpr bool operator()(data_type to)
+  {
+    constexpr auto fromSize = sizeof(FromType);
+    return cudf::type_dispatcher(to, is_logically_castable_to_impl<FromType>{}, fromSize);
+  }
+
+  template <typename FromType,
+            typename std::enable_if_t<cudf::is_fixed_point<FromType>()>* = nullptr>
```

Comment:
Jake's comments from above apply here.  There's no need to SFINAE on `is_fixed_point`.  `device_storage_type_t` will give you the true on-device size of everything.